src/libxively/gateway/xi_gw.c

```diff
@@ -0,0 +1,232 @@
+/* Copyright (c) 2003-2018, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xively_gateway.h>
+#include <xi_types.h>
+#include <xi_gw_layer_stack.h>
+#include <xi_handle.h>
+#include <xi_globals.h>
+#include <xi_context.h>
+#include <xi_gateway_context.h>
+
+#include <xi_vector.h>
+#include <xi_helpers.h>
+
+typedef struct xi_ed_id_ed_context_pair_s
+{
+    char* edge_device_id;
+    xi_context_t* edge_device_context;
+} xi_ed_id_ed_context_pair_t;
+
+static int8_t _xi_ed_find_context_2_id( const union xi_vector_selector_u* e0,
+                                        const union xi_vector_selector_u* e1 )
+{
+    if ( NULL == e0 || NULL == e1 )
+    {
+        return 1;
+    }
+
+    const xi_ed_id_ed_context_pair_t* pair0 =
+        ( xi_ed_id_ed_context_pair_t* )e0->ptr_value;
+    const xi_ed_id_ed_context_pair_t* pair1 =
+        ( xi_ed_id_ed_context_pair_t* )e1->ptr_value;
+
+    if ( NULL == pair0 || NULL == pair1 )
+    {
+        return 1;
+    }
+
+    return strcmp( pair0->edge_device_id, pair1->edge_device_id );
+}
+
+xi_state_t xi_connect_ed( xi_context_handle_t xih,
```

Comment:
Ok, I am cool with it. But what about these?
`xi_ed_connect`
`xi_edge_device_connect`
Again I am also cool with `xi_gw_connect_ed` formers are just shorter.

Reply:
I'm not in love with xi_gw_connect_ed either. I do like the idea of everything starting with xi_gw though, since this is the Xively C Client Gateway module that will be optional. 

how about xi_gw_edge_device_connect?  I know it's longer, but really it shouldn't appear in may parts of the client's  source code :)

Reply:
ok, did the change
```
xi_gw_edge_device_connect
xi_gw_edge_device_disconnect
xi_gw_edge_device_remove
xi_gw_edge_device_publish
```

Reply:
I am not certain though how many and what function will be in final version. I don't like the disconnect+remove. I'd rather have one instead of those two. We'll see the result...